### DISCUSSION TITLE
How to set public IP for flannel to use with `wireguard-native` backend?

### DISCUSSION THREAD

**Discussion by skirsten** [k3s-io/k3s#5911] [Q&A]
After a lot of trial and error the more I try the less I understand...

Here is my setup:
- k3s `v1.24.3+k3s1`
- All nodes have `--node-external-ip` set to their public IP
- I am using `--flannel-backend "wireguard-native"`

With just this, the `sudo wg` `endpoint`s are set to the private IP's.

Now I figured out that I can annotate the node and then **restart k3s** to force flannel to use the public IP:
```sh
kubectl annotate node my-node flannel.alpha.coreos.com/public-ip-overwrite=1.2.3.4
```

Now I would expect only one side of the tunnel to have the correct endpoint, but for some miraculous reason it actually detects the public IP of the other side and both endpoint's are correct.

So this solution seems to be the correct way. Unfortunately I cannot use it, because its a annotation and not a node-label so I cannot reliably set it before flannel expects it (as far as I know).
So now I am searching for a solution to this:
- Is there some other flag / env var that I can use to set the public IP that flannel will use?
- Is there some way to set the annotation on the `agent` command before anything else?

Any help is apprechiated!

Edit:
Also, I know about https://github.com/alvaroaleman/flannel-node-annotator and https://github.com/alekc-go/flannel-fixer but I am pretty sure these also require a restart :/

**Comment by skirsten** [comment]
I created a PR in flannel to configure this via node-labels. Lets see what they think: https://github.com/flannel-io/flannel/pull/1622

**Comment by brandond** [comment]
I was under the impression that it should do the right thing if you set node-external-ip. Have you by any chance disabled the k3s cloud provider? Do you see the correct public IPs in `kubectl get node -o wide`?

**Reply by skirsten** [reply]
Hi,
I think on the old wireguard backend (non-native) it used to get the public-ip from some combination of node-ip and node-external-ip.
But with the new native backend the only thing k3s forwards to flannel is the interface.

And no, I did not disable the cloud-provider. Its still working as expected and setting the ExternalIP.
Flannel does not care about this ExternalIP or the `k3s.io/external-ip` annotation though.
For reference, here is the relevant `describe node` output without the `public-ip-overwrite` annotation:

```
Annotations:        flannel.alpha.coreos.com/backend-data: {"PublicKey":"xxx"}
                    flannel.alpha.coreos.com/backend-type: wireguard
                    flannel.alpha.coreos.com/kube-subnet-manager: true
                    flannel.alpha.coreos.com/public-ip: 10.128.15.197
                    k3s.io/external-ip: 34.171.97.xxx
                    k3s.io/hostname: node-7c5536f
                    k3s.io/internal-ip: 10.128.15.197
                    k3s.io/node-args:
                      ["agent","--server","https://xxx:6443","--node-ip","10.128.15.197","--node-external-ip","34.171.97.xxx","--node-label","infra....
                    ...
Addresses:
                    InternalIP:  10.128.15.197
                    ExternalIP:  34.171.97.xxx
                    Hostname:    node-7c5536f
```

and after adding the `flannel.alpha.coreos.com/public-ip-overwrite: 34.171.97.xxx` annotation and **restarting** k3s on this node, the `flannel.alpha.coreos.com/public-ip` annotation and `sudo wg` output are updated to the provided overwrite.

**Reply by brandond** [reply]
Hmm, that probably needs to be improved on the k3s side then.

**Reply by skirsten** [reply]
For now I just bit the bullet and wrote my own node annotator that syncs the overwrite annotations for flannel (and also kilo which I later added) and removes a taint when the annotations are okay.

**Comment by jamcole** [comment]
Ran across this solution today which solved a problem of mine...

There's now a `--public-ip` option for flannel, if k3s exposed that as `--flannel-public-ip` that would be an ideal solution vs the add annotation and restart approach we have to do now. https://github.com/flannel-io/flannel/blob/master/Documentation/configuration.md

**Reply by brandond** [reply]
There is already a flag for this.

> --flannel-external-ip | Use node external IP addresses as the destination for Flannel traffic, instead of internal IPs. Only applies when --node-external-ip is set on a node.

This is set on servers, and controls flannel behavior on all cluster members.

**Reply by jamcole** [reply]
I saw that option, but it's exactly opposite of what I needed, I needed the public-ip to be set to the node-ip to send all traffic over the ethernet connection vs the wifi connection of a small cluster... Basically, node-ip, not external-ip.

**Reply by brandond** [reply]
What you're asking for doesn't make any sense to me. You want to use the external IP as the internal IP? Kubernetes external IPs are meant to be a public IP that is NATed to the internal private IP. It's not really meant to select between interfaces on a host with multiple interfaces, and probably won't do what you want if you try to use it that way.

**Reply by jamcole** [reply]
Not trying to do that all, I want to use the wireless for ingress and egress traffic for the cluster, and the ethernet for all inter-node communication.

Even with node-ip (eth0) and external-ip (wlan0) set correctly for k3s, all flannel traffic to other nodes was going over the "first" or "default" interface / route, which happens to be the wlan0 (external-ip) interface...

The flannel "public-ip" option isn't equivalent to "external-ip", it's actually:
> IP accessible by other nodes for inter-host communication. see: https://github.com/flannel-io/flannel/blob/master/Documentation/running.md#interface-selection

Basically, even with --flannel-iface specified, flannel was choosing to send all inter-host traffic over the default gateway interface, which is very non-optimal. Whatever autodetection logic is in flannel, it was not correct for this cluster.

The ONLY way I could get flannel to use the correct interface and ip for all inter-node traffic was to apply the annotation
flannel.alpha.coreos.com/public-ip-overwrite: '$node-ip' to each node.

**Reply by brandond** [reply]
Yes, the logic for `--flannel-external-ip` sets `flannel.alpha.coreos.com/public-ip-overwrite` to the node's external IP, if one is set.

> flannel was choosing to send all inter-host traffic over the default gateway interface, which is very non-optimal.

If you set `--flannel-iface`, flannel should advertise the IP associated with that interface in the address annotation, and use that for inter-node CNI traffic. Is that not what you were seeing? Note that interface selection for the purposes of packet routing is still up to the kernel; if the kernel decides to send traffic to the wlan0 interface on another node, out the local eth0 interface, because that is the lowest cost route - that is not a K3s issue.

Note that this only affects CNI traffic; K3s and Kubernetes itself will still use the selected node-ip/node-external-ip values for inter-host communication.
